helm-v3/006-repositories.md

```diff
@@ -1,40 +1,177 @@
-> [name=Matt Fisher] Adnan, would you mind clarifying your comment a little?
-> 
-> [name=Adnan Abdulhussein] Sorry, the proposal mentions there will be a
-> default implemention for `helm push` to a regular HTTP(S) chart repository.
-> It's unclear to me how this would actually work.
-> 
-> [name=Adnan Abdulhussein] Sounds like this is something that will be fleshed
-> out more in Josh's proposal, so I'll wait for that :)
+- A repository v2 specification will be defined, and the index file will be
+JSON (`index.json` vs `index.yaml`)
+
+## Capabilities API
+
+The Capabilties API is designed to provide Helm with information such as:
+
+1. Does this domain support actions such as login or push?
+2. What supported method should the client use to perform a specific action?
+3. What metadata does the client need to initiate the action?
+
+This will allow Helm to support any number of things in the future related
+to repositories without making breaking changes (e.g. remote `helm search`).
+
+### `/.well-known/helm`
+
+The Capabilities API will be enabled for a specific domain via
+[Well-Known URI](https://tools.ietf.org/html/rfc5785) located at
+`/.well-known/helm`.
+
+*Note: This file may also serve to provide configuration for other APIs
+in the future by supplying an additional top-level section.*
+
+The response body should be valid JSON and should supply a valid,
+top-level `capabilities` section. For example:
+
+```
+{
+    "capabilities": {
+        "login": {
+            ...
+        },
+        "push": {
+            ...
+        },
+        "search": {
+            ...
+        }
+    }
+}
+```
+
+Each subsection (i.e. `login`, `push`, `search`) will provide configuration
+specific to the capabiltity.
+
+For example, the following might inform Helm that this domain only
+accepts token-based login:
+
+```
+...
+        "login": {
+            "methods": {
+                "token": {
+                    "loginUrl": "https://auth.site.com/oauth2/authorize",
+                    "tokenUrl": "https://auth.site.com/oauth2/token"
+                }
+            }
+        }
+...
+```
+
+## The `helm login` command
+
+A new `helm login` command will be added to Helm, responsible for
+managing authentication across repository providers.
+
+The basic usage will be:
+
+```
+helm login <url>
+```
+
+This command is intended to allow users to authenticate just one time in
+order to access multiple repositories provided by the same source. 
+
+Afterwards, running a `helm repo add` pointing to any repository URL
+tied to a provider that has been successfully authenticated against will
+not require additional credentials.
+
+The exact semantics are TBD, but several auth methods will be supported,
+such as:
+
+* Basic Auth
+* Cert based Auth
+* Token/Bearer Auth
+
+### Storing Credentials
+
+Credentials obtained via `helm login` will be stored in a new `providers`
+section in `$HELM_HOME/repository/repositories.yaml`.
```

Comment:
Without enraging a yaml/json/xml/ini file debate, might suggest json here, to be consistent with other local config files, and not really intended to be clean/human readable

Reply:
While I agree, this is already an existing file in $HELM_HOME, and to change the format "just because" might not be appropriate.

In terms of consistency, I think we can all agree to at least keep all server-side files in JSON (index.json, /.well-known/helm).

Reply:
Would it be valuable to remove this section? If we remove sections where we describe Helm's technical architecture, the proposal's core focus then describes how a helm repository is architected, while keeping the helm client's technical implementation to a minimum. It also frees up the verbiage from "here's how helm interacts with a repository" to "here's how YOU can write a client that interacts with a repository".

A few reference implementations (`helm login`/`helm push`) to point to are good to have, but optimally this proposal should be able to stand on its own for other clients to write against.

It also keeps the Helm yaml/json/xml/ini debates to a minimum ;)

Reply:
Yes, you're right. The only way to get these proposals merged is to limit scope, we can add this back at a later date if necessary.